src/v/storage/disk_log_impl.cc

```diff
@@ -844,17 +872,28 @@ ss::future<bool> disk_log_impl::sliding_window_compact(
         seg->force_set_commit_offset_from_index();
         seg->release_batch_cache_index();
 
+        const auto removed_bytes = ssize_t(size_before) - ssize_t(size_after);
+
         // Mark the segment as completed window compaction, and possibly set the
         // clean_compact_timestamp in it's index.
-        co_await internal::mark_segment_as_finished_window_compaction(
-          seg, is_clean_compacted, *_probe);
+        bool marked_as_cleanly_compacted
+          = co_await internal::mark_segment_as_finished_window_compaction(
+            seg, is_clean_compacted, *_probe);
+
+        // If marked cleanly compacted, remove the entire seg size (before
+        // compaction) from dirty segment bytes. Otherwise, only deduct removed
+        // bytes.
+        const auto dirty_removed_bytes = marked_as_cleanly_compacted
+                                           ? size_before
+                                           : removed_bytes;
+        update_dirty_segment_bytes<subtract_tag>(dirty_removed_bytes);
```

Comment:
nit: maaaybe more intuitive as

```
update_dirty_segment_bytes<subtract_tag>(size_before);
if (!marked_as_cleanly_compacted) {
    update_dirty_segment_bytes<add_tag>(size_after);
}
```

Reply:
Sure, up to you. I was thinking about it from the perspective of replacement being conceptually equivalent to removal and an addition, and thought it was easier to follow that way.